src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+            lock (_sync)
+            {
+                if (_writingHead != null)
+                {
+                    // Commit the data as not already committed
+                    CommitUnsynchronized();
+                }
+
+                awaitable = _readerAwaitable.Complete();
+
+                cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, _signalWriterAwaitable, this);
+            }
+
+            cancellationTokenRegistration.Dispose();
+
+            TrySchedule(_readerScheduler, awaitable);
+
+            return new PipeAwaiter<FlushResult>(_writer);
+        }
+
+        internal void CompleteWriter(Exception exception)
+        {
+            Action awaitable;
+            PipeCompletionCallbacks completionCallbacks;
+            bool readerCompleted;
+
+            lock (_sync)
+            {
+                if (_currentWriteLength > 0)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_CompleteWriterActiveWriter();
+                }
+
+                completionCallbacks = _writerCompletion.TryComplete(exception);
+                awaitable = _readerAwaitable.Complete();
+                readerCompleted = _readerCompletion.IsCompleted;
+            }
+
+            if (completionCallbacks != null)
+            {
+                TrySchedule(_readerScheduler, _invokeCompletionCallbacks, completionCallbacks);
+            }
+
+            TrySchedule(_readerScheduler, awaitable);
+
+            if (readerCompleted)
+            {
+                CompletePipe();
+            }
+        }
+
+        internal void AdvanceReader(SequencePosition consumed)
+        {
+            AdvanceReader(consumed, consumed);
+        }
+
+        internal void AdvanceReader(SequencePosition consumed, SequencePosition examined)
+        {
+            BufferSegment returnStart = null;
+            BufferSegment returnEnd = null;
+
+            Action continuation = null;
+            lock (_sync)
+            {
+                var examinedEverything = false;
+                if (examined.Segment == _commitHead)
+                {
+                    examinedEverything = _commitHead != null ? examined.Index == _commitHeadIndex - _commitHead.Start : examined.Index == 0;
+                }
+
+                if (consumed.Segment != null)
+                {
+                    if (_readHead == null)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_AdvanceToInvalidCursor();
+                        return;
+                    }
+
+                    var consumedSegment = (BufferSegment)consumed.Segment;
+
+                    returnStart = _readHead;
+                    returnEnd = consumedSegment;
+
+                    // Check if we crossed _maximumSizeLow and complete backpressure
+                    long consumedBytes = new ReadOnlySequence<byte>(returnStart, _readHeadIndex, consumedSegment, consumed.Index).Length;
```

Comment:
Creating a ReadOnlySequence here just to get the length seems wasteful here.

Can we just do something similar to https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs#L246?
```C#
long consumedBytes = (endSegment.RunningIndex - start.Next.RunningIndex) // Length of data in between first and last segment
                   + (start.Memory.Length - startIndex) // Length of data in first segment
                   + endIndex; // Length of data in last segment
```

Reply:
No we can not, no one should interpret Positions except the class that created them.

Reply:
@pakrym what happens if the SequencePosition passed in isn't backed by a BufferSegment? Won't be break anyways?

Reply:
Sooo, awesome question David. The answer is Pipe.cs code is wrong when dealing with Positions, it only works because I happened to select 0, 0 for index prefixes when ROS stores `IMemoryList`, so you can read Indexes unmodified.

I think we can merge this as is for now, but we have to do something with this. Noone should be reading index from SequencePosition except ROS.

I propose adding `MemoryMarshal.TryGetOwnedMemory/MemoryList/Array(ReadOnlySequence<T>, out ...)` to be able to get values from ROS correctly.